Refine config #137

mike820324 opened the pull request:
Hi @chhsiao90, 
finally, I have finished the refine config part, sorry for taking so long. The PR changed the following part.
- add `list:str`, `list:int` instead of using list option type.
- change `string` and `boolean` into `str` and `bool`.
- properly support boolean type.
- add new parameter, config_file_flags to define section and key of config file.
- change the layout proxy config file.

Help me review this a bit, any suggestion would be helpful.

TODO:
- unittest
- setup.py

chhsiao90 commented on microproxy/config.py:
```diff
@@ -167,18 +179,17 @@ def type_transform(config, optionInfo):
-        elif optionInfo[field]["type"] == "list":
+        if "list" == option_type:
+            list_type = optionInfo[field]["type"].split(":")[1]
+            type_func = getattr(__builtin__, list_type)
```
I think it could just define a list of support type, ex: list, str, int, bool etc.
And we only handle for the fix set of type.
I think there is no need for the flexibility that sacrifice some of the readability.

mike820324 replied:
Make sense, I'll push new commit to change this part.

chhsiao90 commented on microproxy/test/test_config.py:
```diff
@@ -99,184 +106,148 @@ def test_list_type_option(self):
+    #     self.assertIsInstance(config["port"], int)
+    #     self.assertEqual(config["port"], 8080)
+    #     self.assertIsInstance(config["http_port"], list)
+    #     self.assertListEqual(config["http_port"], [5000, 5001])
```
If we didn't need this test, then just removed it!

chhsiao90 reviewed (commented):
LGTM~ Just a little comment, please see my feedback~

mike820324 commented:
@chhsiao90 it's all set. Please help me review this if you got some time. If everything is okay, I'll summarize the commits and merge this PR. thanks.

chhsiao90 reviewed (approved):
LGTM~
BTW, I had noticed that the previous commit of master branch is 27 days ago~~

mike820324 commented:
:< sorry about that. I'll push a new release today.

mike820324 committed:
Refine config.py and command line interface.

- Add new option types.
- Properly handle boolean type.
- Change subcommand into individual binary.

mike820324 committed:
Update unittest.

mike820324 committed:
Update setup.py and Makefile.

- Can use pip install to install the different binary interface.
- Add some shortcut in Makefile to install mpxxx in virtualenv.

chhsiao90 commented:
Congratulation~~~~
